File: apple/TriageCompanion/SnykClient.swift
import Foundation

struct SnykClient {
    private let token: String
    private let baseURL: URL
    private let http: HTTPClient

    init(token: String, baseURL: String, http: HTTPClient = HTTPClient()) throws {
        self.token = token
        guard let url = URL(string: try normalizedSnykAPIBaseURL(baseURL)) else {
            throw TriageCompanionError.message("Snyk API base URL is invalid.")
        }
        self.baseURL = url
        self.http = http
    }

    func listOpenIssues() async throws -> [SnykIssueItem] {
        let organizations = try await loadOrganizations()
        var issues: [SnykIssueItem] = []

        for organization in organizations {
            let projectNames = try await loadProjectNames(organizationID: organization.id)
            let records = try await paginate(
                path: "orgs/\(encodedPathComponent(organization.id))/issues",
                query: [
                    "status": "open",
                    "ignored": "false"
                ]
            )
            for record in records {
                issues.append(try issueItem(record, organization: organization, projectNames: projectNames))
            }
        }

        return issues.sorted {
            let leftRank = severityRank($0.severity)
            let rightRank = severityRank($1.severity)
            if leftRank != rightRank {
                return leftRank > rightRank
            }
            if $0.organizationName != $1.organizationName {
                return $0.organizationName < $1.organizationName
            }
            if $0.projectName != $1.projectName {
                return $0.projectName < $1.projectName
            }
            return $0.updatedAt > $1.updatedAt
        }
    }

    private var headers: [String: String] {
        [
            "Authorization": "Token \(token)",
            "Accept": "application/vnd.api+json",
            "User-Agent": "triage-companion",
            "Cache-Control": "no-cache",
            "Pragma": "no-cache"
        ]
    }

    private func loadOrganizations() async throws -> [SnykOrganization] {
        let records = try await paginate(path: "orgs", query: [:])
        var organizations: [SnykOrganization] = []
        var seen: Set<String> = []

        for record in records {
            let id = try apiPathID(string(record, key: "id", context: "Snyk organization"), label: "Snyk organization ID")
            let attributes = try object(record, key: "attributes", context: "Snyk organization \(id)")
            let slug = try apiPathID(string(attributes, key: "slug", context: "Snyk organization \(id)"), label: "Snyk organization slug")
            let name = try string(attributes, key: "name", context: "Snyk organization \(id)")
            if seen.insert(id).inserted {
                organizations.append(SnykOrganization(id: id, slug: slug, name: name))
            }
        }

        return organizations
    }

    private func loadProjectNames(organizationID: String) async throws -> [String: String] {
        let records = try await paginate(
            path: "orgs/\(encodedPathComponent(organizationID))/projects",
            query: [:]
        )
        var names: [String: String] = [:]
        for record in records {
            let id = try apiPathID(string(record, key: "id", context: "Snyk project"), label: "Snyk project ID")
            let attributes = try object(record, key: "attributes", context: "Snyk project \(id)")
            names[id] = try firstString(attributes, keys: ["name", "target_reference", "origin"], context: "Snyk project \(id)")
        }

        return names
    }

    private func issueItem(
        _ record: [String: Any],
        organization: SnykOrganization,
        projectNames: [String: String]
    ) throws -> SnykIssueItem {
        let issueID = try apiPathID(string(record, key: "id", context: "Snyk issue"), label: "Snyk issue ID")
        let attributes = try object(record, key: "attributes", context: "Snyk issue \(issueID)")
        let projectID = try projectID(from: record, context: "Snyk issue \(issueID)")
        let projectName = projectNames[projectID] ?? "Unavailable project \(projectID)"
        let severity = try string(attributes, key: "effective_severity_level", context: "Snyk issue \(issueID)")
        guard severityRank(severity) > 0 else {
            throw TriageCompanionError.message("Snyk issue \(issueID) included an unknown severity.")
        }
        let status = try string(attributes, key: "status", context: "Snyk issue \(issueID)")
        guard status.lowercased() == "open" else {
            throw TriageCompanionError.message("Snyk issue \(issueID) was not open.")
        }
        if let ignored = attributes["ignored"] as? Bool, ignored {
            throw TriageCompanionError.message("Snyk issue \(issueID) was ignored.")
        }

        let title = try string(attributes, key: "title", context: "Snyk issue \(issueID)")
        let issueType = try string(attributes, key: "type", context: "Snyk issue \(issueID)")
        let issueKey = try string(attributes, key: "key", context: "Snyk issue \(issueID)")
        let updatedAt = try parseServiceDate(
            string(attributes, key: "updated_at", context: "Snyk issue \(issueID)"),
            label: "Snyk issue updated_at"
        )
        guard let url = snykIssueURL(organizationSlug: organization.slug, projectID: projectID, issueKey: issueKey) else {
            throw TriageCompanionError.message("Could not build Snyk issue URL.")
        }

        return SnykIssueItem(
            id: "\(organization.id)#\(issueID)",
            organizationName: organization.name,
            projectName: projectName,
            packageName: packageName(from: attributes),
            severity: severity,
            issueType: issueType,
            title: title,
            updatedAt: updatedAt,
            url: url
        )
    }

    private func paginate(path: String, query: [String: String]) async throws -> [[String: Any]] {
        var queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        queryItems.append(URLQueryItem(name: "version", value: "2024-10-15"))
        queryItems.append(URLQueryItem(name: "limit", value: "100"))

        guard let routeURL = URL(string: "\(baseURL.absoluteString)/\(path)") else {
            throw TriageCompanionError.message("Snyk API route was invalid.")
        }
        var url = urlWithQuery(routeURL, queryItems: queryItems)
        var seen = Set<String>()
        try recordSnykPaginationURL(&seen, url)
        var records: [[String: Any]] = []

        while true {
            let (payload, _) = try await http.jsonObject(
                from: request(url: url, headers: headers),
                serviceName: "Snyk"
            )
            guard let data = payload["data"] as? [[String: Any]] else {
                throw TriageCompanionError.message("Snyk response must include a data array.")
            }
            records.append(contentsOf: data)

            guard let links = payload["links"] as? [String: Any], let next = try paginationHref(from: links["next"]) else {
                break
            }
            let nextURL = try resolvedPaginationURL(next, currentURL: url)
            if data.isEmpty {
                throw TriageCompanionError.message("Snyk response returned an empty page before pagination finished.")
            }
            try recordSnykPaginationURL(&seen, nextURL)
            url = nextURL
        }

        return records
    }

    private func paginationHref(from value: Any?) throws -> String? {
        guard let value else {
            return nil
        }
        if let string = value as? String {
            return try cleanRequiredText(string, label: "Snyk pagination link")
        }
        if let object = value as? [String: Any], let href = object["href"] as? String {
            return try cleanRequiredText(href, label: "Snyk pagination link")
        }

        throw TriageCompanionError.message("Snyk pagination link must be a URL string.")
    }

    private func resolvedPaginationURL(_ value: String, currentURL: URL) throws -> URL {
        guard rawPathSegments(value) != nil else {
            throw TriageCompanionError.message("Snyk pagination link must stay on the current API route.")
        }

        let resolved: URL?
        if value.hasPrefix("/") && !value.hasPrefix("//") {
            resolved = URL(string: "\(baseURL.absoluteString)\(value)")
        } else {
            resolved = URL(string: value, relativeTo: currentURL)?.absoluteURL
        }

        guard let resolved else {
            throw TriageCompanionError.message("Snyk pagination link must be a valid URL.")
        }

        return try validatedPaginationURL(resolved, currentURL: currentURL)
    }

    private func validatedPaginationURL(_ value: URL, currentURL: URL) throws -> URL {
        guard value.user == nil, value.password == nil else {
            throw TriageCompanionError.message("Snyk pagination link must not include credentials.")
        }
        guard value.port == nil else {
            throw TriageCompanionError.message("Snyk pagination link must not include a port.")
        }
        guard value.fragment == nil else {
            throw TriageCompanionError.message("Snyk pagination link must not include fragments.")
        }
        guard isSupportedSnykAPIURL(value) else {
            throw TriageCompanionError.message("Snyk pagination link must stay on a US-hosted REST API base URL.")
        }
        guard sameOrigin(value, currentURL),
              rawPathSegments(value.absoluteString) == rawPathSegments(currentURL.absoluteString) else {
            throw TriageCompanionError.message("Snyk pagination link must stay on the current API route.")
        }

        guard let version = queryValue(value, name: "version") else {
            throw TriageCompanionError.message("Snyk pagination link must include a REST API version.")
        }
        guard version == queryValue(currentURL, name: "version") else {
            throw TriageCompanionError.message("Snyk pagination link must keep the current REST API version.")
        }
        guard stableSnykPaginationQuery(value) == stableSnykPaginationQuery(currentURL) else {
            throw TriageCompanionError.message("Snyk pagination link must keep the current API query.")
        }

        return value
    }

    private func recordSnykPaginationURL(_ seen: inout Set<String>, _ url: URL) throws {
        let key = snykPaginationLoopKey(url)
        guard seen.insert(key).inserted else {
            throw TriageCompanionError.message("Snyk pagination repeated a previously fetched page.")
        }
    }

    private func snykPaginationLoopKey(_ url: URL) -> String {
        normalizedURLString(url, excluding: [])
    }

    private func stableSnykPaginationQuery(_ url: URL) -> String {
        sortedQuery(url, excluding: ["page", "starting_after", "ending_before"])
    }

    private func isSupportedSnykAPIURL(_ url: URL) -> Bool {
        [defaultSnykAPIBaseURL, alternateSnykAPIBaseURL].contains { value in
            guard let allowedURL = URL(string: value), sameOrigin(url, allowedURL) else {
                return false
            }
            let path = url.path
            let allowedPath = allowedURL.path
            return path == allowedPath || path.hasPrefix("\(allowedPath)/")
        }
    }

    private func sameOrigin(_ left: URL, _ right: URL) -> Bool {
        left.scheme?.lowercased() == right.scheme?.lowercased()
            && left.host?.lowercased() == right.host?.lowercased()
            && left.port == right.port
    }

    private func rawPathSegments(_ value: String) -> [String]? {
        let pathAndSuffix: String
        if let schemeRange = value.range(of: "//") {
            let afterAuthority = value[schemeRange.upperBound...]
            if let pathStart = afterAuthority.firstIndex(of: "/") {
                pathAndSuffix = String(afterAuthority[pathStart...])
            } else {
                pathAndSuffix = "/"
            }
        } else if value.hasPrefix("?") || value.hasPrefix("#") || value.isEmpty {
            pathAndSuffix = ""
        } else {
            pathAndSuffix = value
        }

        let pathEnd = pathAndSuffix.indices.first { index in
            pathAndSuffix[index] == "?" || pathAndSuffix[index] == "#"
        } ?? pathAndSuffix.endIndex
        let path = String(pathAndSuffix[..<pathEnd])
        if path.isEmpty {
            return []
        }

        let parts = path.split(separator: "/", omittingEmptySubsequences: false)
        let hasLeadingSlash = parts.first == ""
        let hasTrailingSlash = parts.last == ""
        let segments = hasLeadingSlash
            ? (hasTrailingSlash ? parts.dropFirst().dropLast() : parts.dropFirst())
            : (hasTrailingSlash ? parts.dropLast() : parts[...])
        if segments.contains(where: { $0.isEmpty }) {
            return nil
        }

        var decoded: [String] = []
        for segment in segments {
            guard let value = String(segment).removingPercentEncoding, value != ".", value != ".." else {
                return nil
            }
            decoded.append(value)
        }

        return decoded
    }

    private func queryValue(_ url: URL, name: String) -> String? {
        URLComponents(url: url, resolvingAgainstBaseURL: false)?
            .queryItems?
            .first { $0.name == name }?
            .value
    }

    private func normalizedURLString(_ url: URL, excluding excludedNames: Set<String>) -> String {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return url.absoluteString
        }
        let queryItems = components.queryItems?
            .filter { !excludedNames.contains($0.name) }
            .sorted {
                if $0.name == $1.name {
                    return ($0.value ?? "") < ($1.value ?? "")
                }
                return $0.name < $1.name
            }
        components.queryItems = queryItems?.isEmpty == false ? queryItems : nil

        return components.url?.absoluteString ?? url.absoluteString
    }

    private func sortedQuery(_ url: URL, excluding excludedNames: Set<String>) -> String {
        guard let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems else {
            return ""
        }

        return queryItems
            .filter { !excludedNames.contains($0.name) }
            .sorted {
                if $0.name == $1.name {
                    return ($0.value ?? "") < ($1.value ?? "")
                }
                return $0.name < $1.name
            }
            .map { item in
                if let value = item.value {
                    return "\(encodedQueryComponent(item.name))=\(encodedQueryComponent(value))"
                }
                return encodedQueryComponent(item.name)
            }
            .joined(separator: "&")
    }

    private func encodedQueryComponent(_ value: String) -> String {
        var allowed = CharacterSet.urlQueryAllowed
        allowed.remove(charactersIn: "&=+")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    private func object(_ value: [String: Any], key: String, context: String) throws -> [String: Any] {
        guard let object = value[key] as? [String: Any] else {
            throw TriageCompanionError.message("\(context) must include \(key).")
        }

        return object
    }

    private func string(_ value: [String: Any], key: String, context: String) throws -> String {
        guard let stringValue = value[key] as? String else {
            throw TriageCompanionError.message("\(context) must include \(key).")
        }

        return try cleanRequiredText(stringValue, label: "\(context) \(key)")
    }

    private func firstString(_ value: [String: Any], keys: [String], context: String) throws -> String {
        for key in keys {
            if let stringValue = value[key] as? String, !stringValue.isEmpty {
                return try cleanRequiredText(stringValue, label: "\(context) \(key)")
            }
        }

        throw TriageCompanionError.message("\(context) must include one of \(keys.joined(separator: ", ")).")
    }

    private func apiPathID(_ value: String, label: String) throws -> String {
        guard value.range(of: #"^[A-Za-z0-9._-]+$"#, options: .regularExpression) != nil, value != ".", value != ".." else {
            throw TriageCompanionError.message("\(label) must be a safe API path segment.")
        }

        return value
    }

    private func projectID(from record: [String: Any], context: String) throws -> String {
        let relationships = try object(record, key: "relationships", context: context)
        let scanItem = try object(relationships, key: "scan_item", context: context)
        let data = try object(scanItem, key: "data", context: context)
        guard let type = data["type"] as? String, type == "project" else {
            throw TriageCompanionError.message("\(context) scan_item relationship must point at a project.")
        }

        return try apiPathID(string(data, key: "id", context: context), label: "Snyk project ID")
    }

    private func packageName(from attributes: [String: Any]) -> String? {
        guard let coordinates = attributes["coordinates"] as? [[String: Any]] else {
            return nil
        }
        for coordinate in coordinates {
            guard let representations = coordinate["representations"] as? [[String: Any]] else {
                continue
            }
            for representation in representations {
                if let package = representation["package"] as? [String: Any],
                   let name = package["name"] as? String,
                   !name.isEmpty {
                    return name
                }
            }
        }

        return nil
    }

    private func snykIssueURL(organizationSlug: String, projectID: String, issueKey: String) -> URL? {
        let appOrigin = baseURL.absoluteString == alternateSnykAPIBaseURL ? "https://app.us.snyk.io" : "https://app.snyk.io"
        return URL(
            string: "\(appOrigin)/org/\(encodedPathComponent(organizationSlug))/project/\(encodedPathComponent(projectID))#issue-\(encodedPathComponent(issueKey))"
        )
    }
}

private struct SnykOrganization {
    let id: String
    let slug: String
    let name: String
}

File: android/app/src/main/kotlin/com/triagecompanion/app/SnykClient.kt
package com.triagecompanion.app

import java.net.URI
import java.net.URISyntaxException
import org.json.JSONObject

class SnykClient(
    token: String,
    baseURL: String,
    private val http: HttpClient = HttpClient(),
) {
    private val baseURL: String = normalizedSnykAPIBaseURL(baseURL)
    private val headers = mapOf(
        "Authorization" to "Token $token",
        "Accept" to "application/vnd.api+json",
        "User-Agent" to "triage-companion",
        "Cache-Control" to "no-cache",
        "Pragma" to "no-cache",
    )

    private class SnykOrganization(val id: String, val slug: String, val name: String)

    suspend fun listOpenIssues(): List<SnykIssueItem> {
        val organizations = loadOrganizations()
        val issues = mutableListOf<SnykIssueItem>()

        for (organization in organizations) {
            val projectNames = loadProjectNames(organization.id)
            val records = paginate(
                path = "orgs/${encodedPathComponent(organization.id)}/issues",
                query = listOf(
                    "status" to "open",
                    "ignored" to "false",
                ),
            )
            for (record in records) {
                issues.add(issueItem(record, organization, projectNames))
            }
        }

        return issues.sortedWith { left, right ->
            val leftRank = severityRank(left.severity)
            val rightRank = severityRank(right.severity)
            when {
                leftRank != rightRank -> rightRank.compareTo(leftRank)
                left.organizationName != right.organizationName ->
                    left.organizationName.compareTo(right.organizationName)
                left.projectName != right.projectName -> left.projectName.compareTo(right.projectName)
                else -> right.updatedAt.compareTo(left.updatedAt)
            }
        }
    }

    private suspend fun loadOrganizations(): List<SnykOrganization> {
        val records = paginate(path = "orgs", query = emptyList())
        val organizations = mutableListOf<SnykOrganization>()
        val seen = mutableSetOf<String>()

        for (record in records) {
            val id = apiPathID(string(record, "id", context = "Snyk organization"), label = "Snyk organization ID")
            val attributes = obj(record, "attributes", context = "Snyk organization $id")
            val slug = apiPathID(
                string(attributes, "slug", context = "Snyk organization $id"),
                label = "Snyk organization slug",
            )
            val name = string(attributes, "name", context = "Snyk organization $id")
            if (seen.add(id)) {
                organizations.add(SnykOrganization(id, slug, name))
            }
        }

        return organizations
    }

    private suspend fun loadProjectNames(organizationID: String): Map<String, String> {
        val records = paginate(
            path = "orgs/${encodedPathComponent(organizationID)}/projects",
            query = emptyList(),
        )
        val names = mutableMapOf<String, String>()
        for (record in records) {
            val id = apiPathID(string(record, "id", context = "Snyk project"), label = "Snyk project ID")
            val attributes = obj(record, "attributes", context = "Snyk project $id")
            names[id] = firstString(
                attributes,
                keys = listOf("name", "target_reference", "origin"),
                context = "Snyk project $id",
            )
        }

        return names
    }

    private fun issueItem(
        record: JSONObject,
        organization: SnykOrganization,
        projectNames: Map<String, String>,
    ): SnykIssueItem {
        val issueID = apiPathID(string(record, "id", context = "Snyk issue"), label = "Snyk issue ID")
        val attributes = obj(record, "attributes", context = "Snyk issue $issueID")
        val projectID = projectID(record, context = "Snyk issue $issueID")
        val projectName = projectNames[projectID] ?: "Unavailable project $projectID"
        val severity = string(attributes, "effective_severity_level", context = "Snyk issue $issueID")
        if (severityRank(severity) <= 0) {
            throw TriageCompanionException("Snyk issue $issueID included an unknown severity.")
        }
        val status = string(attributes, "status", context = "Snyk issue $issueID")
        if (status.lowercase() != "open") {
            throw TriageCompanionException("Snyk issue $issueID was not open.")
        }
        if (attributes.opt("ignored") == true) {
            throw TriageCompanionException("Snyk issue $issueID was ignored.")
        }

        val title = string(attributes, "title", context = "Snyk issue $issueID")
        val issueType = string(attributes, "type", context = "Snyk issue $issueID")
        val issueKey = string(attributes, "key", context = "Snyk issue $issueID")
        val updatedAt = parseServiceDate(
            string(attributes, "updated_at", context = "Snyk issue $issueID"),
            label = "Snyk issue updated_at",
        )

        return SnykIssueItem(
            id = "${organization.id}#$issueID",
            organizationName = organization.name,
            projectName = projectName,
            packageName = packageName(attributes),
            severity = severity,
            issueType = issueType,
            title = title,
            updatedAt = updatedAt,
            url = snykIssueURL(organization.slug, projectID, issueKey),
        )
    }

    private suspend fun paginate(path: String, query: List<Pair<String, String>>): List<JSONObject> {
        val queryItems = query + listOf(
            "version" to "2024-10-15",
            "limit" to "100",
        )
        var url = urlWithQuery("$baseURL/$path", queryItems)
        val seen = mutableSetOf<String>()
        recordPaginationURL(seen, url)
        val records = mutableListOf<JSONObject>()

        while (true) {
            val (payload, _) = http.jsonObject(url, headers, serviceName = "Snyk")
            val data = payload.optJSONArray("data")
                ?: throw TriageCompanionException("Snyk response must include a data array.")
            for (index in 0 until data.length()) {
                val record = data.opt(index) as? JSONObject
                    ?: throw TriageCompanionException("Snyk response must include a data array.")
                records.add(record)
            }

            val links = payload.optJSONObject("links")
            val next = links?.let { paginationHref(if (it.isNull("next")) null else it.opt("next")) } ?: break
            val nextPageURL = resolvedPaginationURL(next, currentURL = url)
            if (data.length() == 0) {
                throw TriageCompanionException("Snyk response returned an empty page before pagination finished.")
            }
            recordPaginationURL(seen, nextPageURL)
            url = nextPageURL
        }

        return records
    }

    private fun paginationHref(value: Any?): String? {
        if (value == null) {
            return null
        }
        if (value is String) {
            return cleanRequiredText(value, label = "Snyk pagination link")
        }
        if (value is JSONObject) {
            val href = value.opt("href") as? String
            if (href != null) {
                return cleanRequiredText(href, label = "Snyk pagination link")
            }
        }

        throw TriageCompanionException("Snyk pagination link must be a URL string.")
    }

    private fun resolvedPaginationURL(value: String, currentURL: String): String {
        if (rawPathSegments(value) == null) {
            throw TriageCompanionException("Snyk pagination link must stay on the current API route.")
        }

        val resolved: URI? = if (value.startsWith("/") && !value.startsWith("//")) {
            parsedURI("$baseURL$value")
        } else {
            parsedURI(currentURL)?.let { current ->
                try {
                    current.resolve(value)
                } catch (_: IllegalArgumentException) {
                    null
                }
            }
        }
        if (resolved == null) {
            throw TriageCompanionException("Snyk pagination link must be a valid URL.")
        }

        return validatedPaginationURL(resolved, currentURL).toString()
    }

    private fun validatedPaginationURL(value: URI, currentURL: String): URI {
        val current = parsedURI(currentURL)
            ?: throw TriageCompanionException("Snyk pagination link must be a valid URL.")
        if (value.userInfo != null) {
            throw TriageCompanionException("Snyk pagination link must not include credentials.")
        }
        if (value.port != -1) {
            throw TriageCompanionException("Snyk pagination link must not include a port.")
        }
        if (value.rawFragment != null) {
            throw TriageCompanionException("Snyk pagination link must not include fragments.")
        }
        if (!isSupportedSnykAPIURL(value)) {
            throw TriageCompanionException("Snyk pagination link must stay on a US-hosted REST API base URL.")
        }
        if (!sameOrigin(value, current) ||
            rawPathSegments(value.toString()) != rawPathSegments(current.toString())
        ) {
            throw TriageCompanionException("Snyk pagination link must stay on the current API route.")
        }

        val version = queryValue(value, "version")
            ?: throw TriageCompanionException("Snyk pagination link must include a REST API version.")
        if (version != queryValue(current, "version")) {
            throw TriageCompanionException("Snyk pagination link must keep the current REST API version.")
        }
        if (stablePaginationQuery(value) != stablePaginationQuery(current)) {
            throw TriageCompanionException("Snyk pagination link must keep the current API query.")
        }

        return value
    }

    private fun recordPaginationURL(seen: MutableSet<String>, url: String) {
        val parsed = parsedURI(url)
            ?: throw TriageCompanionException("Snyk pagination link must be a valid URL.")
        if (!seen.add(paginationLoopKey(parsed))) {
            throw TriageCompanionException("Snyk pagination repeated a previously fetched page.")
        }
    }

    private fun paginationLoopKey(url: URI): String {
        val query = sortedQueryString(url, excluding = emptySet())
        val base = "${url.scheme}://${url.host}${url.rawPath ?: ""}"
        return if (query.isEmpty()) base else "$base?$query"
    }

    private fun stablePaginationQuery(url: URI): String =
        sortedQueryString(url, excluding = setOf("page", "starting_after", "ending_before"))

    private fun isSupportedSnykAPIURL(url: URI): Boolean =
        listOf(DEFAULT_SNYK_API_BASE_URL, ALTERNATE_SNYK_API_BASE_URL).any { allowed ->
            val allowedURL = parsedURI(allowed) ?: return@any false
            if (!sameOrigin(url, allowedURL)) {
                return@any false
            }
            val path = url.path ?: ""
            val allowedPath = allowedURL.path ?: ""
            path == allowedPath || path.startsWith("$allowedPath/")
        }

    private fun sameOrigin(left: URI, right: URI): Boolean =
        left.scheme?.lowercase() == right.scheme?.lowercase() &&
            left.host?.lowercase() == right.host?.lowercase() &&
            left.port == right.port

    private fun rawPathSegments(value: String): List<String>? {
        val pathAndSuffix: String
        val authorityIndex = value.indexOf("//")
        if (authorityIndex >= 0) {
            val afterAuthority = value.substring(authorityIndex + 2)
            val pathStart = afterAuthority.indexOf('/')
            pathAndSuffix = if (pathStart >= 0) afterAuthority.substring(pathStart) else "/"
        } else if (value.startsWith("?") || value.startsWith("#") || value.isEmpty()) {
            pathAndSuffix = ""
        } else {
            pathAndSuffix = value
        }

        val pathEnd = pathAndSuffix.indexOfFirst { it == '?' || it == '#' }
        val path = if (pathEnd >= 0) pathAndSuffix.substring(0, pathEnd) else pathAndSuffix
        if (path.isEmpty()) {
            return emptyList()
        }

        val parts = path.split("/")
        val hasLeadingSlash = parts.firstOrNull() == ""
        val hasTrailingSlash = parts.size > 1 && parts.last() == ""
        val start = if (hasLeadingSlash) 1 else 0
        val end = if (hasTrailingSlash) parts.size - 1 else parts.size
        val segments = if (start <= end) parts.subList(start, end) else emptyList()
        if (segments.any { it.isEmpty() }) {
            return null
        }

        val decoded = mutableListOf<String>()
        for (segment in segments) {
            val decodedSegment = percentDecoded(segment)
            if (decodedSegment == null || decodedSegment == "." || decodedSegment == "..") {
                return null
            }
            decoded.add(decodedSegment)
        }

        return decoded
    }

    private fun obj(value: JSONObject, key: String, context: String): JSONObject =
        value.optJSONObject(key)
            ?: throw TriageCompanionException("$context must include $key.")

    private fun string(value: JSONObject, key: String, context: String): String {
        val stringValue = value.opt(key) as? String
            ?: throw TriageCompanionException("$context must include $key.")

        return cleanRequiredText(stringValue, label = "$context $key")
    }

    private fun firstString(value: JSONObject, keys: List<String>, context: String): String {
        for (key in keys) {
            val stringValue = value.opt(key) as? String
            if (stringValue != null && stringValue.isNotEmpty()) {
                return cleanRequiredText(stringValue, label = "$context $key")
            }
        }

        throw TriageCompanionException("$context must include one of ${keys.joinToString(", ")}.")
    }

    private fun apiPathID(value: String, label: String): String {
        if (!value.matches(Regex("^[A-Za-z0-9._-]+$")) || value == "." || value == "..") {
            throw TriageCompanionException("$label must be a safe API path segment.")
        }

        return value
    }

    private fun projectID(record: JSONObject, context: String): String {
        val relationships = obj(record, "relationships", context)
        val scanItem = obj(relationships, "scan_item", context)
        val data = obj(scanItem, "data", context)
        if (data.opt("type") as? String != "project") {
            throw TriageCompanionException("$context scan_item relationship must point at a project.")
        }

        return apiPathID(string(data, "id", context), label = "Snyk project ID")
    }

    private fun packageName(attributes: JSONObject): String? {
        val coordinates = attributes.optJSONArray("coordinates") ?: return null
        for (coordinateIndex in 0 until coordinates.length()) {
            val coordinate = coordinates.opt(coordinateIndex) as? JSONObject ?: continue
            val representations = coordinate.optJSONArray("representations") ?: continue
            for (representationIndex in 0 until representations.length()) {
                val representation = representations.opt(representationIndex) as? JSONObject ?: continue
                val name = representation.optJSONObject("package")?.opt("name") as? String
                if (name != null && name.isNotEmpty()) {
                    return name
                }
            }
        }

        return null
    }

    private fun snykIssueURL(organizationSlug: String, projectID: String, issueKey: String): String {
        val appOrigin = if (baseURL == ALTERNATE_SNYK_API_BASE_URL) "https://app.us.snyk.io" else "https://app.snyk.io"
        return "$appOrigin/org/${encodedPathComponent(organizationSlug)}/project/${encodedPathComponent(projectID)}#issue-${encodedPathComponent(issueKey)}"
    }

    private fun parsedURI(value: String): URI? =
        try {
            URI(value)
        } catch (_: URISyntaxException) {
            null
        }
}
